Improve `hover` and `find-references` output

usfbih8u opened the pull request:
Changes:

 - Add line content and replace absolute with relative path in `showLocations()`. Instead of showing only the absolute path of the reference found, this patch adds the line content with `getLineContentFromPath()`. This new function will return a string with the content of the line or, in case of error, an error message. Also replaces the absolute path with the relative path in order to  shorten it and leave more space for the line content. For this task  `relPathFromFileUri()` was implemented.

 - Add `showHoverInfo()` to create an horizontal split with hover results.

usfbih8u commented:
The motivation behind this PR is to make `hover` and `find-references` more useful.

As stated in https://github.com/Andriamanitra/mlsp/issues/7 `hover` feature is not as useful as it could be (for example `clangd` can respond with multiple lines only of annotations that are in the code as documentation). I think as @r3t4k3r, that `hsplit` is the best option until micro can expose some tooltip implementation.

Also this PR gives more context to the respond of `find-references` printing the line where the reference its located.

Let me know what you think.

Andriamanitra reviewed (changes requested):
Thanks for the contribution! I apologize for the delayed response.

I have not made `hover` open a new split because it then needs to be manually closed, which is usually not what I would personally want. But you're probably right that it should be changed to a hsplit (at least until we get tooltip support in micro itself) so it's actually possible to read longer doc strings. That part could be merged.



However I'm not so sure about the changes to `find-references` output.
I prefer absolute paths because they are more fool proof (you can copy-paste them anywhere without thinking about the current working directory). Relative paths help a bit with the length issue when the project root is deeply nested but even relative paths are often too long to be shown at the beginning of the line (the label is the more interesting part). Including a preview of the line content seems reasonable though.

Another consideration is that the filepath should be easily "greppable" to play nice with other plugins. I have been using [another plugin](https://github.com/Andriamanitra/dotfiles/blob/main/micro/.config/micro/plug/ag2/ag.lua) that includes functions to open the found results when they're formatted just right (so you can [bind Enter to quickly jump to the relevant location](https://github.com/Andriamanitra/dotfiles/blob/main/micro/.config/micro/bindings.json#L81-L82)). The output format of `showLocations` (used by both `find-references` and `document-symbols`) was designed with this in mind – the label goes first so it's visible even if the path that follows is very long, and using a character that's unlikely to occur in a filename as the separator (`'\t'`) makes it trivially easy to parse [^1].
I've been meaning to include something like the `OpenResult*` functions from the other plugin in `mlsp` too but just haven't gotten around to it yet.

[^1]: Paths like `/foo/bar:1:2 /baz qux:3:4` are technically possible – POSIX allows file paths to contain spaces and colons. When the `fpath` is at the *end* of a line we know that the last numbers must refer to the row and column. Having `fpath` in the beginning with arbitrary content after (like in your proposed format) on the other hand makes it rather tricky or even impossible to grep the part of the line that represents the file path.

usfbih8u commented on main.lua:
```diff
@@ -277,6 +278,13 @@ function LSPClient:handleResponseError(method, error)
     end
 end
 
+function showHoverInfo(results)
```
I did want to inline it inside [method == "textDocument/hover" branch](https://github.com/usfbih8u/mlsp/blob/95183098e322d2a591ed7c383d8e7045e8a9adb7/main.lua#L306C5-L306C47) but did not know your preferences. Now its in there (this way luacheck does not scream about undefined functions).

usfbih8u commented on main.lua:
```diff
@@ -277,6 +278,13 @@ function LSPClient:handleResponseError(method, error)
+function showHoverInfo(results)
+    local bf = buffer.NewBuffer(results, "[µlsp] hover")
+    bf.Type.Scratch = true
+    bf.Type.Readonly = true
```
In my mind makes more sense to not change the data provided by the LSP (and you always can copy it and change it in the destination buffer). But if you prefer to been able to change it, I am okey with that.

Tell me what you want and I will change it.

Andriamanitra replied:
Either way is fine. I don't think there is any harm in allowing users to edit it if they so choose but it's more of a theoretical use case anyway.

usfbih8u commented:
About the absolute vs relative path. I think its a trade-off. Or absolute path or relative and line content. As you said even relative path is sometimes a large text.

The format I choose was the same as grep output (widely used and supported by other programs). And similar as you do, I use [micro-quickfix](https://github.com/serge-v/micro-quickfix) to jump to the locations (it expects the file at the beginning). I suppose all the tab thing its because the format that uses your `ag` plugin (?). 

And you are right about the label being first but i wanted to keep the same format in both.

That being said, I think that the right call its, as you said, to make the LSP responsible of that and do not leave it to other plugins.

I think using something like your `OpenResult*()` and specific parsers by buffer name will be a nice solution. This way we can use formats that makes more sense for each specific LSP result and do not be constraint.

I think something like:

```lua
-- this functions will return a string (ex: my_project/src/main.lua:33:22)
local parsersByName = {
    ["references"] = function () ... end,
    ["document symbols"] = function () ... end,
    -- other buffers if needed 
}

-- method (string): open, tab, hsplit ...
-- close (boolean)
function jumpToLocation(bufpane, method, close)
    local parser = parsersByName[bp:Name()]
    if not parser then infobar:Error(); return end
    local line = ... -- get the line where the main cursor is
    local destination = parser(line)
    bufpane:HandleCommand(method .." ".. destination)
    if close then
        -- close buffer opened by LSP
    end
end
```

Then we can use two approaches:

- Expose wrappers for each specific case as I recall from your plugin or,
- Expose only `jumpToLocation()` and use the `args` to pass `method` and `close` (verifying them inside) in the binding. 

Example: "command:jumpToLocation open true" (or we can use directly "close" instead of true so its more explicit at what it does, no second argument means false).

Maybe use another name less generic if its exposed, like `lspJump` or something with LSP in the name.

Andriamanitra commented:
> That being said, I think that the right call its, as you said, to make the LSP responsible of that and do not leave it to other plugins.
> 
> I think using something like your OpenResult*() and specific parsers by buffer name will be a nice solution. This way we can use formats that makes more sense for each specific LSP result and do not be constraint.

I agree that's a good solution.
For `find-references` we should probably simply print the location on its own line? That way there is no need to worry about parsing weird edge cases and we can still have the preview. If the output is marked read-only the `jumpToLocation` (or whatever it will be called) command could even be smart enough to look for the location on the previous line if the current line number mod 3 is 1. And we could eventually even have custom file types to syntax highlight the outputs.

I propose the following formatting:

## `document-symbols`
```
[Method] (*BufPane).ScrollUp          # /tmp/micro/internal/action/actions.go:25:1
[Method] (*BufPane).ScrollDown        # /tmp/micro/internal/action/actions.go:32:1
[Method] (*BufPane).ScrollAdjust      # /tmp/micro/internal/action/actions.go:40:1
[Method] (*BufPane).MousePress        # /tmp/micro/internal/action/actions.go:51:1
[Method] (*BufPane).MouseDrag         # /tmp/micro/internal/action/actions.go:103:1
[Method] (*BufPane).MouseRelease      # /tmp/micro/internal/action/actions.go:124:1
[Method] (*BufPane).ScrollUpAction    # /tmp/micro/internal/action/actions.go:143:1
[Method] (*BufPane).ScrollDownAction  # /tmp/micro/internal/action/actions.go:149:1
```

## `find-references`
```
/tmp/micro/internal/display/softwrap.go:63:2
	Scroll(s SLoc, n int) SLoc

/tmp/micro/internal/action/actions.go:27:18
	v.StartLine = h.Scroll(v.StartLine, -n)

/tmp/micro/internal/action/actions.go:34:18
	v.StartLine = h.Scroll(v.StartLine, n)

/tmp/micro/internal/action/actions.go:44:19
		v.StartLine = h.Scroll(end, -h.BufView().Height+1)

/tmp/micro/internal/action/actions.go:157:18
	v.StartLine = h.Scroll(h.SLocFromLoc(h.Cursor.Loc), -h.BufView().Height/2)

/tmp/micro/internal/action/actions.go:174:14
		SLoc:    h.Scroll(v.StartLine, scrollmargin),

/tmp/micro/internal/action/actions.go:185:14
		SLoc:    h.Scroll(v.StartLine, h.BufView().Height/2),

/tmp/micro/internal/action/actions.go:198:16
	lastLine := h.Scroll(v.StartLine, h.BufView().Height-1)
```

usfbih8u commented:
The format for symbols looks nice and the alignment should be easy to implement. The absolute path could make the line wrap but maybe we could set softwrap to false for that specific buffer.

But I do not like the find references one, it is weird to read and uses a lot of space. I think best solution is relative path + line content (and as stated before we can set softwrap to false in the buffer if necessary). 

It is there any good reason for using absolute path over relative? What is your use case?

And notice that your `absPathFromFileUri()` function returns whatever uri the LSP returns (I catch a bug in my relative function, see [issue](https://github.com/Andriamanitra/mlsp/issues/11), because it returned a relative path instead, see [commit](https://github.com/usfbih8u/mlsp/commit/95183098e322d2a591ed7c383d8e7045e8a9adb7))

About the syntax I use a special one for the references, quickfix, etc that puts the file and location in other color so its easy to read but expects to be as grep output.

Similar to this:

```yaml
filetype: lsp-syntax

detect:
    filename: "(^references$)|(^document symbols$)"

rules:
    - identifier: "^[^[:space:]][^:]+:[0-9]+:[0-9]+:"
    - identifier: "^[^[:space:]][^:]+:[0-9]+:"
    - identifier: "^[^[:space:]][^:]+:"
```

We could use a color for each type of label. But I do not know a lot about the syntax part of micro.

usfbih8u commented:
Now with https://github.com/Andriamanitra/mlsp/pull/14 can I assume that `absPathFromFileUri()` always returns absolute path?

No need for https://github.com/Andriamanitra/mlsp/pull/10/commits/95183098e322d2a591ed7c383d8e7045e8a9adb7?

Andriamanitra commented:
> Now with #14 can I assume that `absPathFromFileUri()` always returns absolute path?
> 
> No need for [9518309](https://github.com/Andriamanitra/mlsp/commit/95183098e322d2a591ed7c383d8e7045e8a9adb7)?

Yeah – it always did when it was given a file URI. Problem is that despite the name `lspLocations[1].uri` is **not** always a file URI, it can also be just a path (without "file://" prefix). That path used to be relative but now it's absolute so you should be able to remove that commit.

But the original implementation has a bug too. It won't work if the file is not under the current working directory. For example if you're editing `/etc/hosts` and your working directory is `/home/user/` the relative path should be `../../etc/hosts`, but `relPathFromFileUri` returns `.`. This is part of the reason I prefer absolute paths: they're much easier to reason about than relative paths. If we go with relative paths I think we should use [`filepath.Rel`](https://pkg.go.dev/path/filepath#Rel) to (hopefully) correctly handle all the edge cases.

usfbih8u commented:
You are right, I did not see `filepath.Rel` 😅.

I did not think about the plugin handling files from '/' being inside '$HOME'... this makes me think. It's the plugin sending changes to the LSP of files that are not related to that LSP?

For example, I am editing a markdown or python file with the lua LSP activated, the plugin handles that? Is the plugin sending `didOpen` messages (or others) for that files?

This PRs are my first contact with LSPs, so i dont know much. Can the LSP respond with the file types it does not handles?

Andriamanitra commented:
> I did not think about the plugin handling files from '/' being inside '$HOME'... this makes me think. It's the plugin sending changes to the LSP of files that are not related to that LSP?
> 
> For example, I am editing a markdown or python file with the lua LSP activated, the plugin handles that? Is the plugin sending `didOpen` messages (or others) for that files?
> 
> This PRs are my first contact with LSPs, so i dont know much. Can the LSP respond with the file types it does not handles?

Yes, µlsp (LSP client) currently tells the language server about all open files regardless of file type or path. It is up to the language server to decide what (if anything) to do with that. This is somewhat wasteful if you're editing Python files with a Lua language server running, but I don't think it is much of a problem in practice.

Eventually it might be nice to have support for [dynamicRegistration](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#client_registerCapability) which would allow the language servers to tell the client which requests it supports and for which files. Then we would not need to send textDocument/didOpen and textDocument/didChange messages for files that the language server is not interested in.

usfbih8u commented:
I normally use other tabs for documentation, notes, etc. 

I've been reading about dynamicRegistration, watching what is the result of initialize (i saw the languageId part of the request from the client) and searching in the specs. It seems that the LSPs do not respond directly with what kind of files are supported (at least the ones i tried). Is this correct? Can you point me in the right direction if i am missing something?

Did you read https://github.com/Andriamanitra/mlsp/pull/10#issuecomment-2450296697 ? Is this PR completed?

Andriamanitra commented:
> I've been reading about dynamicRegistration, watching what is the result of initialize (i saw the languageId part of the request from the client) and searching in the specs. It seems that the LSPs do not respond directly with what kind of files are supported (at least the ones i tried). Is this correct? Can you point me in the right direction if i am missing something?

I have not coded it yet so I'm not entirely certain about the details but if I understand the spec correctly it should work something like this:
1. Client opts in by setting `capabilities.fileOperations.dynamicRegistration = true` in the `initialize` request
2. Server may send `client/registerCapability` requests to the client. These may **optionally** include a list of [`DocumentFilter`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentFilter) to pick which documents the capability applies to.
3. The client can choose to not make requests that there is no registered capability for.

Although now that I read more about it looks like the `initialize` response could also contain document filters in `ServerCapabilities`. None of the language servers I tried so far (pylsp, ruff, rust-analyzer) seem to take advantage of that though. Not much point in supporting the feature if none of the servers use it.
 
> Did you read [#10 (comment)](https://github.com/Andriamanitra/mlsp/pull/10#issuecomment-2450296697) ? Is this PR completed?

Yes I read it, but I still think we should go with the format I proposed in https://github.com/Andriamanitra/mlsp/pull/10#issuecomment-2448534992. The current format in this PR is IMO
1. Less readable for the labeled case because the labels are after the locations
2. Ambiguous in the presence of spaces and colons (eg. `/foo/bar:1:2:/baz qux:3:4 abc` could be either `path="/foo/bar:1:2:/baz /qux", line=3, column=4, content="abc"` or `path="/foo/bar:1:2", line=1, column=2, content="/baz /qux:3:4:abc"`)
3. Less useful because of the use of relative paths: you can't use them in eg. a file explorer or a terminal that has a different working directory than micro.

If you have a strong argument for relative paths I could be convinced otherwise, but points 1 and 2 need to be addressed before this can be merged.

Andriamanitra commented:
I'm also a bit worried about the performance of `showLocations` if the references are deep into a file and there are many of them. Imagine something like `/project/main.c:90001`, `/project/main.c:90003`, `/project/main.c:90004` – it would read over 90000 lines from the `example.c` file 3 times! It's not that unrealistic of a scenario either as there are often many references to the same variable in the same file.

usfbih8u commented:
I read that part diagonally and understood that it is the client's responsibility to set a filter on the server. For example, using `clang`, it should not respond to hover requests for `*.h` files, only for `*.c` files. This could improve server performance, but it does not address the issue of sending requests from the client, does it?

Maybe rely on the buffer's file type or use the same approach you implemented for the `lsp` command (which I haven't read yet). Perhaps a combination of both could work?

About the format:

> 1. Less readable for the labeled case because the labels are after the locations

The format for symbols is ok to me, as I said before.

> 2. Ambiguous in the presence of spaces and colons (eg. /foo/bar:1:2:/baz qux:3:4 abc could be either path="/foo/bar:1:2:/baz /qux", line=3, column=4, content="abc" or path="/foo/bar:1:2", line=1, column=2, content="/baz /qux:3:4:abc")

Fair enough, I guess.

> 3. Less useful because of the use of relative paths: you can't use them in eg. a file explorer or a terminal that has a different working directory than micro.

I do not know if that is the normal use case. micro's "Shell bar" uses current directory, same micros's terminal, new tabs in terminal emulators use the current directory as well, same as tmux (and similar apps) splits and windows.

> If you have a strong argument for relative paths I could be convinced otherwise, but points 1 and 2 need to be addressed before this can be merged.

My biggest concerns:

1. The name length in the tab bar displays the entire path, making it unusable (even two tabs are too much). This is the main reason that I avoid absolute paths when opening a buffer (this applies to both formats).
2. The proposed format for find-references is, in my opinion, hard to read and takes up too much space. Could this next format be a middle ground?

```references
main.lua:
    22: local res = nil
    33: if res then

other.lua:
    15: local res = req
```

- **Pros**: similar to yours, avoids ambiguity with spaces and is more compact.
- **Cons**: getting the location is not immediate (no mod 3 operation; we have to search backwards line by line until we encounter first line with first char not equal to tab, but should be fast).

About the performance of `showLocations()`:

This was good enough to make the PR and see if you are ok with the main idea, if not i would keep like it is (i can not use micro with a file that large lol). The last format make us to sort the result first (**are the results already sorted by the LSP?**), so after all entries are sorted we can open and read the file only once.

Andriamanitra commented:
> I read that part diagonally and understood that it is the client's responsibility to set a filter on the server. For example, using `clang`, it should not respond to hover requests for `*.h` files, only for `*.c` files. This could improve server performance, but it does not address the issue of sending requests from the client, does it?

The client would not send requests if the server hasn't advertised it has the capability to handle it. [This part is already implemented in µlsp](https://github.com/Andriamanitra/mlsp/blob/79ba8f28840c83ce373e7a58c6daf3770e742f1e/main.lua#L1180-L1195), although it's missing support for the document filtering. Although the requests related to document synchronization (`textDocument/didOpen`, `textDocument/didChange`, etc.) do not go through it because according to the spec every language server must support them.
 
> Maybe rely on the buffer's file type or use the same approach you implemented for the `lsp` command (which I haven't read yet). Perhaps a combination of both could work?

From what I've seen it seems language servers don't give a document filter for these requests either, so it's impossible to know (reliably) which file types a given server is interested in. It would be possible to hard code their supported file types but I don't really like that approach as it's brittle and maintenance-heavy. It is not always obvious which file types a language server should be associated with. For example language servers might care about changes to configuration files like Cargo.toml or package.json in addition to source code files.

> > If you have a strong argument for relative paths I could be convinced otherwise, but points 1 and 2 need to be addressed before this can be merged.
> 
> My biggest concerns:
> 
>     1. The name length in the tab bar displays the entire path, making it unusable (even two tabs are too much). This is the main reason that I avoid absolute paths when opening a buffer (this applies to both formats).

We could do the conversion from absolute path to relative path when opening the file when that is implemented.
 
>     2. The proposed format for find-references is, in my opinion, hard to read and takes up too much space. Could this next format be a middle ground?
> 
> 
> ```
> main.lua:
>     22: local res = nil
>     33: if res then
> 
> other.lua:
>     15: local res = req
> ```
> 
>     * **Pros**: similar to yours, avoids ambiguity with spaces and is more compact.
> 
>     * **Cons**: getting the location is not immediate (no mod 3 operation; we have to search backwards line by line until we encounter first line with first char not equal to tab, but should be fast).

I'm fine with this format. The implementation is a bit more complicated but I think it looks better than my suggestion. With this format you will probably want to replace `showLocations` with two separate functions as the outputs of `find-references` and `document-symbols` will not share much any more.

> The last format make us to sort the result first (**are the results already sorted by the LSP?**), so after all entries are sorted we can open and read the file only once.

The specification does not guarantee that they are sorted but in practice I think they will be (for `find-references`). It wouldn't hurt to sort them anyway just to be safe.

usfbih8u commented:
> The client would not send requests if the server hasn't advertised it has the capability to handle it. [This part is already implemented in µlsp](https://github.com/Andriamanitra/mlsp/blob/79ba8f28840c83ce373e7a58c6daf3770e742f1e/main.lua#L1180-L1195), although it's missing support for the document filtering. Although the requests related to document synchronization (`textDocument/didOpen`, `textDocument/didChange`, etc.) do not go through it because according to the spec every language server must support them.

Yes, a filter for a valid capability. The problem is in the client, not in the server (now that I use the plugin a little bit, the delay when opening multiple buffers simultaneously is very noticeable).

> From what I've seen it seems language servers don't give a document filter for these requests, so it's impossible to know (reliably) which file types a given server is interested in. 

I have seen the same, it is strange that something like that is not in the spec.

> It would be possible to hard code their supported file types but I don't really like that approach as it's brittle and maintenance-heavy. It is not always obvious which file types a language server should be associated with. For example language servers might care about changes to configuration files like Cargo.toml or package.json in addition to source code files.

If the LSP uses other types we can not rely in micro filetype for that. But it is already necessary to hardcode the LSP in `config.lua`. We could implement an option that, when not `nil`, restricts the available filetypes/extensions for the client. So nothing has to change but there is the possibility to use it.

> I'm fine with this format. The implementation is a bit more complicated but I think it looks better than my suggestion. With this format you will probably want to replace `showLocations` with two separate functions as the outputs of `find-references` and `document-symbols` will not share much any more.
>
> The specification does not guarantee that they are sorted but in practice I think they will be (for `find-references`). It wouldn't hurt to sort them anyway just to be safe.

Both formats are already completed, except for minor issues (some trouble with '\t' in symbols, and references are not sorted yet, although they were for the LSPs that I tried)

The trouble with the tabs is the visual representation

```
[String]	VERSION                         # main.lua:1:1
[Variable]	micro                         # main.lua:3:7
```

tab is counted as 1 using `#labels[i]` and `util.CharacterCountInString(labels[i])` but when represented you can see there are 4 or 2 spaces. I have to check what micro does.

Andriamanitra commented:
> Yes, a filter for a valid capability. The problem is in the client, not in the server (now that I use the plugin a little bit, the delay when opening multiple buffers simultaneously is very noticeable).

Number of buffers should not affect anything. It is normal for opening a buffer to take time to read the entire file & send its contents to the language server so it might not be instant for larger files. However if it gets slower as you open more buffers that is likely a bug. Please open a new issue if that is the case.

usfbih8u commented:
No, it is not a bug, I was referring to the actual behavior. But does not matter, it is not critical.

usfbih8u commented:
I did not solve the '\t' issue. There is a function in Micro, `StringWidth()`, in `internal/util/util.go` that addresses it.

I added a possible TODO to highlight the reference search in the new buffer created.

I also changed to absolute paths, so they can later be replaced with relative paths during the "jump."

Andriamanitra commented on main.lua:
```diff
@@ -465,7 +473,7 @@ function LSPClient:handleResponseResult(method, result)
             end
             table.insert(symbolLabels, string.format("[%s]\t%s", SYMBOLKINDS[sym.kind], sym.name))
```
This should fix the alignment:
```suggestion
            table.insert(symbolLabels, string.format("%-15s %s", "["..SYMBOLKINDS[sym.kind].."]", sym.name))
```

usfbih8u replied:
I thought those labels were directly from the LSP. I replaced the tab with the size that it occupies, but these are fewer operations. I will use this instead.

Andriamanitra commented on main.lua:
```diff
@@ -1234,22 +1251,103 @@ end
+        }
+    end
+
+    table.sort(references, function(a, b)
+        if a.path < b.path then return true
+        elseif a.path > b.path then return false
+        else
+            if a.line < b.line then return true
+            elseif a.line > b.line then return false
+            else
+                if a.column < b.column then return true
+                elseif a.column > b.column then return false
+                else return true end
+            end
```
This can be simplified:
```suggestion
    table.sort(references, function(a, b)
        if a.path ~= b.path then return a.path < b.path end
        if a.line ~= b.line then return a.line < b.line end
        return a.column < b.column
```

Andriamanitra commented on main.lua:
```diff
@@ -1234,22 +1251,103 @@ end
-        bufContents = bufContents .. line
+    end)
+
+    local bufContents = ""
+    local curFile = nil
+    local file = nil
+    local lineCount = 0
+    local prevLine = 0
+    local lineContent
+    for _, ref in ipairs(references) do
+        repeat -- fake-goto to use break as continue
+            if curFile == ref.path and prevLine == ref.line then break -- Same line? add it directly
+            elseif curFile ~= ref.path then --Handle file opening and closing
+                if file then file:close() end
+                curFile = ref.path
+                if bufContents ~= "" then bufContents = bufContents .. "\n" end
+                bufContents = bufContents .. curFile .. ":\n"
+                lineCount = 0
+                file = io.open(curFile, "rb")
+                if not file then
+                    lineContent = ""
+                    break
+                end
+            end
+
+            repeat
+                lineContent = file:read("*l")
+                lineCount = lineCount + 1
+            until not lineContent or lineCount == ref.line
+            ---Empty string: line not found
+            lineContent = lineContent and lineContent or ""
+
+        until true
+        bufContents = bufContents .. string.format("\t%d:%d:%s\n",ref.line, ref.column, lineContent)
+        prevLine = ref.line
     end
 
+    if file then file:close() end -- last iteration does not close last file
+
     local newBuffer = buffer.NewBuffer(bufContents, newBufferTitle)
```
I think this is too complex, it's hard to follow so deeply nested logic. This suggestion also changes it to collect the lines into an array to avoid repeated string concatenation, which improves readability and should be way more efficient too when there are lots of references.
```suggestion
    local bufLines = {}
    local curFilePath = ""
    local file = nil
    local lineCount = 0
    local lineContent = ""
    for _, ref in ipairs(references) do
        if curFilePath ~= ref.path then
            if file then file:close() end
            if #bufLines > 0 then table.insert(bufLines, "") end
            curFilePath = ref.path
            table.insert(bufLines, curFilePath)
            file = io.open(curFilePath, "rb")
            lineCount = 0
        end

        -- file can be nil if io.open failed
        if file ~= nil then
            while lineCount < ref.line do
                lineContent = file:read("*l")
                lineCount = lineCount + 1
            end
        end
        table.insert(bufLines, string.format("\t%d:%d:%s", ref.line, ref.column, lineContent or ""))
    end

    if file then file:close() end -- last iteration does not close last file
    table.insert(bufLines, "")

    local newBuffer = buffer.NewBuffer(table.concat(bufLines, "\n"), newBufferTitle)
```

Andriamanitra reviewed (changes requested):
The functionality seems good, I just have a couple of suggestions to simplify the implementation a bit.
